phpBB/phpbb/forum/controller/viewforum.php

```diff
@@ -0,0 +1,248 @@
+		return $this->render_forum($forum_id, $parameters);
+	}
+
+	/**
+	 * Renders the forum.
+	 *
+	 * @param int		$forum_id	The ID of the forum to display.
+	 * @param string	$parameters	URL parameters.
+	 *
+	 * @return \Symfony\Component\HttpFoundation\Response HTTP response with the rendered page.
+	 */
+	protected function render_forum($forum_id, $parameters)
+	{
+		$forum_data = [];
+		$forum_parents = [];
+		$parameters = $this->parameters;
+		$forum_tracking_info = $active_forum_ary = $forum_rows = $subforums = $valid_categories = $forum_ids_moderator = $moderators = [];
+
+		// @todo: replace with param processing
+		$has_page_param = false;
+
+		try
+		{
+			$forum_data = $this->forum_retriever->get_forum_metadata($forum_id);
+			$this->user->setup('viewforum', $forum_data['forum_style']);
+			$this->visibility_helper->check($forum_data, $parameters);
+
+			if ($this->request->is_set('e', request_interface::GET) && !$this->user->data['is_registered'])
+			{
+				throw new login_required_exception('LOGIN_NOTIFY_FORUM');
+			}
+
+			// Is this forum a link? ... User got here either because the
+			// number of clicks is being tracked or they guessed the id
+			if ($forum_data['forum_type'] == FORUM_LINK && $forum_data['forum_link'])
+			{
+				// Does it have click tracking enabled?
+				if ($forum_data['forum_flags'] & FORUM_FLAG_LINK_TRACK)
+				{
+					$this->forum_retriever->update_link_count($forum_data);
+				}
+
+				// We redirect to the url. The third parameter indicates that external redirects are allowed.
+				redirect($forum_data['forum_link'], false, true);
+			}
+
+			$forum_parents = $this->forum_retriever->get_forum_parents($forum_data);
+
+			// This is mostly done for BC reasons...
+			$forum_data['forum_parents'] = serialize($forum_parents);
+
+			$forum_parents = $this->visibility_helper->filter_forum_parents($forum_parents);
+
+			if ($forum_data['left_id'] != $forum_data['right_id'] - 1)
+			{
+				$rows = $this->forum_retriever->get_subforums($forum_data);
+				$rows = $this->visibility_helper->filter_subforums($rows);
+
+				list($forum_rows, $subforums, $valid_categories, $forum_ids_moderator, $forum_tracking_info) = $this->forum_retriever->get_subforum_hierarchy(
+					$forum_data,
+					$rows
+				);
+
+				$active_forum_ary = $this->forum_retriever->get_active_topic_array($forum_data, $rows);
+				unset ($rows);
+			}
+
+			// @todo: handle this madness
+			//$template->assign_var('S_HAS_SUBFORUM', false);
+			if ($this->config['load_moderators'])
+			{
+				get_moderators($moderators, $forum_id);
+			}
+		}
+		catch (forum_not_found_exception $e)
+		{
+			throw new http_exception(404, $e->getMessage());
+		}
+		catch (forum_password_needed_exception $e)
+		{
+			$password = (isset($parameters['forum_password'])) ? $parameters['forum_password'] : false;
+			return $this->render_helper->render_forum_password_box($forum_data, $password);
+		}
+		catch (login_required_exception $e)
+		{
+			// @todo: login box.
```

Comment:
I can't remember when and what we really did but I worked on something a long time ago to be a sort of login_box drop in replacement in controllers.
Maybe @VSEphpbb can help?